### ISSUE TITLE
Decouple from Triton

### ISSUE THREAD

**Issue by richardburleigh** [state-spaces/mamba#33]
Congratulations team! This is a seriously impressive model.

Thank you also for sharing the weights and supporting the open-source community!

FYI - I was able to successfully fine-tune the 2.8B model and completely demolish squad, showing strong reasoning abilities.

I'm wanting to see how far we can push the inference speed of this model, but the triton dependency is preventing loading it on CPU or doing an ONNX conversion.

Do you have any advice to decouple this dependency or alternative approaches?

Any help would be appreciated!

**Comment by tridao** [comment]
Great to hear about the Squad result!

I don't have much experience with CPU or ONNX deployment. I mostly work with GPUs.
The model architecture is quite different from what's commonly used (Transformers), so there's probably quite a bit to do to make it compatible with the ecosystem.
Some of the operations are implemented in CUDA (selective scan), idk if that would work with ONNX.
The causal_conv1d has a CUDA implementation but you can also just call pytorch.
Triton is used for RMS norm I think, and you can replace that with a reference Pytorch [implementation](https://github.com/state-spaces/mamba/blob/0131c1e94a46fc9f70bcfc9d57962963bb2f0b9e/mamba_ssm/ops/triton/layernorm.py#L35).

**Comment by maedoc** [comment]
I was also interested in a CPU implementation; I was curious if the functions suffixed with `_ref` are reference implementations (like this one https://github.com/state-spaces/mamba/blob/main/mamba_ssm/ops/selective_scan_interface.py#L311) that can be used to guide new implementations?

In the past I had quite some success porting CUDA/OpenCL kernels to CPU using ISPC (which has very similar programming model) but I saw that C10 & CUB are used here and I don't know them very well.

**Comment by kroggen** [comment]
Modified version to work on CPU is available on [this fork](https://github.com/kroggen/mamba-cpu)

I chopped a lot of code, but it works

**Comment by Liyuansongdsb** [comment]
Congratulations! But I have a question that why the 130M model doesn't preform well in training, only 15% accuracy. Am I doing something wrong?

**Comment by liuyamjune** [comment]
Thanks for your contribution @kroggen ! I used your  mamba-cpu, and it works correctly. But converting it to ONNX took really long time. And do you have any suggestions for this?

**Comment by kroggen** [comment]
@liuyamjune I have no knowledge on ONNX

Have you seen if there is already some port of Mamba/Mamba2 to ONNX?
